File: memories/server/controllers/posts.ts
import express, { Request, Response } from 'express';
import mongoose from 'mongoose';

import PostMessage from '../models/postMessage'; // Assuming this is already in TypeScript

const router = express.Router();

// Get all posts
export const getPosts = async (req: Request, res: Response): Promise<void> => { 
    try {
        const postMessages = await PostMessage.find();
        res.status(200).json(postMessages);
    } catch (error) {
        res.status(404).json({ message: (error as Error).message });
    }
};

// Get a single post by ID
export const getPost = async (req: Request, res: Response): Promise<void> => { 
    const { id } = req.params;

    try {
        const post = await PostMessage.findById(id);
        if (!post) return res.status(404).json({ message: "Post not found" });
        res.status(200).json(post);
    } catch (error) {
        res.status(404).json({ message: (error as Error).message });
    }
};

// Create a new post
export const createPost = async (req: Request, res: Response): Promise<void> => {
    const { title, message, selectedFile, creator, tags } = req.body;

    const newPostMessage = new PostMessage({ title, message, selectedFile, creator, tags });

    try {
        await newPostMessage.save();
        res.status(201).json(newPostMessage);
    } catch (error) {
        res.status(409).json({ message: (error as Error).message });
    }
};

// Update an existing post
export const updatePost = async (req: Request, res: Response): Promise<void> => {
    const { id } = req.params;
    const { title, message, creator, selectedFile, tags } = req.body;
    
    if (!mongoose.Types.ObjectId.isValid(id)) return res.status(404).send(`No post with id: ${id}`);

    const updatedPost = { creator, title, message, tags, selectedFile, _id: id };

    try {
        const result = await PostMessage.findByIdAndUpdate(id, updatedPost, { new: true });
        if (!result) return res.status(404).json({ message: "Post not found" });
        res.json(result);
    } catch (error) {
        res.status(409).json({ message: (error as Error).message });
    }
};

// Delete a post
export const deletePost = async (req: Request, res: Response): Promise<void> => {
    const { id } = req.params;

    if (!mongoose.Types.ObjectId.isValid(id)) return res.status(404).send(`No post with id: ${id}`);

    try {
        const result = await PostMessage.findByIdAndRemove(id);
        if (!result) return res.status(404).json({ message: "Post not found" });
        res.json({ message: "Post deleted successfully." });
    } catch (error) {
        res.status(409).json({ message: (error as Error).message });
    }
};

// Like a post
export const likePost = async (req: Request, res: Response): Promise<void> => {
    const { id } = req.params;

    if (!mongoose.Types.ObjectId.isValid(id)) return res.status(404).send(`No post with id: ${id}`);
    
    try {
        const post = await PostMessage.findById(id);
        if (!post) return res.status(404).json({ message: "Post not found" });

        const updatedPost = await PostMessage.findByIdAndUpdate(id, { likeCount: post.likeCount + 1 }, { new: true });
        res.json(updatedPost);
    } catch (error) {
        res.status(409).json({ message: (error as Error).message });
    }
};

export default router;

File: memories/server/controllers/posts.go
package controllers

import (
	"context"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"your_project/models"  // Import the models package
)

// GetPosts retrieves all posts
func GetPosts(c *gin.Context, postCollection *mongo.Collection) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	cursor, err := postCollection.Find(ctx, bson.M{})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	defer cursor.Close(ctx)

	var posts []models.Post  // Use the Post model
	if err = cursor.All(ctx, &posts); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, posts)
}

// GetPost retrieves a single post by its ID
func GetPost(c *gin.Context, postCollection *mongo.Collection) {
	id := c.Param("id")
	objID, _ := primitive.ObjectIDFromHex(id)

	var post models.Post  // Use the Post model
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	err := postCollection.FindOne(ctx, bson.M{"_id": objID}).Decode(&post)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Post not found"})
		return
	}

	c.JSON(http.StatusOK, post)
}

// CreatePost creates a new post
func CreatePost(c *gin.Context, postCollection *mongo.Collection) {
	var post models.Post  // Use the Post model
	if err := c.ShouldBindJSON(&post); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	post.ID = primitive.NewObjectID()
	post.CreatedAt = time.Now()
	post.LikeCount = 0

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	_, err := postCollection.InsertOne(ctx, post)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, post)
}

// UpdatePost updates an existing post
func UpdatePost(c *gin.Context, postCollection *mongo.Collection) {
	id := c.Param("id")
	objID, _ := primitive.ObjectIDFromHex(id)

	var updatedPost models.Post  // Use the Post model
	if err := c.ShouldBindJSON(&updatedPost); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	result, err := postCollection.UpdateOne(ctx, bson.M{"_id": objID}, bson.M{
		"$set": updatedPost,
	})
	if err != nil || result.MatchedCount == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Post not found"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Post updated successfully"})
}

// DeletePost deletes a post by its ID
func DeletePost(c *gin.Context, postCollection *mongo.Collection) {
	id := c.Param("id")
	objID, _ := primitive.ObjectIDFromHex(id)

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	result, err := postCollection.DeleteOne(ctx, bson.M{"_id": objID})
	if err != nil || result.DeletedCount == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Post not found"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Post deleted successfully"})
}

// LikePost increases the like count of a post
func LikePost(c *gin.Context, postCollection *mongo.Collection) {
	id := c.Param("id")
	objID, _ := primitive.ObjectIDFromHex(id)

	var post models.Post  // Use the Post model
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	err := postCollection.FindOne(ctx, bson.M{"_id": objID}).Decode(&post)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Post not found"})
		return
	}

	_, err = postCollection.UpdateOne(ctx, bson.M{"_id": objID}, bson.M{
		"$inc": bson.M{"likeCount": 1},
	})
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Post liked successfully"})
}
